netkan/netkan/mirrorer.py

```diff
@@ -0,0 +1,294 @@
+import boto3
+import internetarchive
+import tempfile
+import urllib.request
+import requests
+import hashlib
+import logging
+from pathlib   import Path
+from .metadata import Ckan
+
+class Mirrorer:
+
+    CACHE_PATH = Path('/home/netkan/ckan_cache/')
+
+    def __init__(self, queue, timeout, ckan_meta_repo, ia_access, ia_secret, ia_collection):
+        sqs = boto3.resource('sqs')
+        self.sqs_client     = boto3.client('sqs')
+        self.queue          = sqs.get_queue_by_name(QueueName=queue)
+        self.timeout        = timeout
+        self.ckan_meta_repo = ckan_meta_repo
+        self.ia_collection  = ia_collection
+        self.ia_session     = internetarchive.get_session(config = {
+            's3': {
+                'access': ia_access,
+                'secret': ia_secret,
+            }
+        })
+
+    def run(self):
+        while True:
+            messages = self.queue.receive_messages(
+                MaxNumberOfMessages   = 10,
+                MessageAttributeNames = ['All'],
+                VisibilityTimeout     = self.timeout,
+            )
+            if messages:
+                # Check if archive.org is overloaded
+                if self.ia_overloaded():
+                    logging.info('The Internet Archive is overloaded, try again later')
+                    continue
+                # Make sure we have the up to date copy of the metadata for the files we're mirroring
+                logging.info('Updating repo')
+                self.ckan_meta_repo.remotes.origin.pull('master', strategy_option='ours')
+                # Start processing the messages
+                to_delete = []
+                for msg in messages:
+                    path = Path(self.ckan_meta_repo.working_dir, msg.body)
+                    if self.try_mirror(CkanMirror(path)):
+                        # Successfully handled -> OK to delete
+                        to_delete.append(self.deletion_msg(msg))
+                self.queue.delete_messages(Entries = to_delete)
+
+    def try_mirror(self, ckan):
+        if not ckan.can_mirror:
+            # If we can't mirror, then we're done with this message
+            logging.info(f'Ckan {ckan.mirror_item} cannot be mirrored')
+            return True
+        if ckan.mirrored(self.ia_session):
+            # If it's already mirrored, then we're done with this message
+            logging.info(f'Ckan {ckan.mirror_item} is already mirrored')
+            return True
+        cached_file = self.cache_find_file(ckan)
+        if cached_file:
+            # If the download is in the cache, use it
+            with open(cached_file, mode='rb') as f:
+                logging.info(f'Found matching cache entry at {cached_file}')
```

Comment:
I don't want `try_upload` to open the file because in its other usage, we already have an open file handle courtesy of `NamedTemporaryFile` (see below).

I'll switch this block to use `Path.open()` though.